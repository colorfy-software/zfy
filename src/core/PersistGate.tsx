import React, { useEffect } from 'react'

import type { CreateStoreType } from '../types'

import useRehydrate from './use-rehydrate'
import { validatePersistGate } from '../internals/validation'

interface PropsType<StoresType> {
  children: JSX.Element | JSX.Element[] | (() => JSX.Element)
  loader?: JSX.Element | (() => JSX.Element)
  stores: StoresType
}

/**
 * Component that renders a loader while all the provided persisted stores get rehydrated.
 * @prop stores - `Record<string, CreateStoreType<any>>`— Object containing all the zustand stores to rehydrate.
 * @prop children - `JSX.Element | JSX.Element[]`— React element to render once rehydration is completed.
 * @prop loader - `JSX.Element`— Optional. Loader component to display.
 */
const PersistGate = <StoresType extends Record<string, CreateStoreType<any>>>({
  stores,
  loader,
  children,
}: PropsType<StoresType>): JSX.Element | null => {
  useEffect(() => {
    validatePersistGate(stores)
  }, [stores])

  const isRehydrated = useRehydrate<StoresType>(stores)

  return (
    <>
      {isRehydrated
        ? typeof children === 'function'
          ? children()
          : children
        : typeof loader === 'function'
        ? loader()
        : loader || null}
    </>
  )
}

export default PersistGate
