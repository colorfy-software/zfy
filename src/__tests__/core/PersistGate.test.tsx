import React from 'react'
import { render, waitFor } from '@testing-library/react-native'

import {
  data,
  SyncStorage,
  rehydratedData,
  assertStoreContent,
} from '../helpers'
import initZfy from '../../core/init-zfy'
import PersistGate from '../../core/PersistGate'
import createStore from '../../core/create-store'
import { Text, View } from 'react-native'

describe('ðŸš§ Core > PersistGate:', () => {
  beforeAll(() => {
    initZfy({
      storage: SyncStorage,
      persistKey: 'appPersistKey',
    })
  })

  it('renders correctly', async () => {
    const loader = jest.fn()
    const store = createStore('jest', data)

    const { toJSON } = await waitFor(() =>
      render(
        <PersistGate stores={{ jest: store }} loader={loader}>
          <View>
            <Text>My App</Text>
          </View>
        </PersistGate>
      )
    )

    expect(toJSON()).toMatchSnapshot()
    expect(loader).toHaveBeenCalledTimes(1)

    expect.assertions(2)
  })

  it('renders loader during rehydration then renders children', async () => {
    const loader = jest.fn()
    const children = jest.fn()

    const store = createStore('jest', data)

    await waitFor(() => {
      render(
        <PersistGate stores={{ jest: store }} loader={loader}>
          {children}
        </PersistGate>
      )

      expect(loader).toHaveBeenCalledTimes(1)
      expect(children).not.toHaveBeenCalled()
      assertStoreContent({ store, expectedData: data })
    })

    expect(loader).toHaveBeenCalledTimes(1)
    expect(children).toHaveBeenCalledTimes(1)
    assertStoreContent({ store, expectedData: rehydratedData })

    expect.assertions(14)
  })
})
