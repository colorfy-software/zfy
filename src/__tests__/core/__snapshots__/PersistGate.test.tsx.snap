// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ğŸš§ Core > PersistGate: renders correctly 1`] = `
<View>
  <Text>
    My App
  </Text>
</View>
`;
